adding ipython interpreter option for developers

vsoch opened the pull request:
This is a small pull request to add an option to use an ipython interpreter for developers, which is my preference (and maybe others too?) Basically, you can specify it like this:

```bash
$ spack python -i ipython
Python 3.8.3 (default, May 19 2020, 18:47:26) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.17.0 -- An enhanced Interactive Python. Type '?' for help.


Spack version 0.16.0
Python 3.8.3, Linux x86_64

In [1]:
```
I replaced the console.runsource with a simple exec, since embed() essentially includes all the locals anyway. For running commands/files I force the user to use python, because while it works to use ipython, it doesn't make sense for a non-interactive use case. 

This is my first PR in a long time so I suspect the formatting will be wrong, I'll follow up with fixes when the CI tells me I'm a wet noodle.

vsoch committed:
adding ipython interpreter option for developers

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch committed:
fixing flake8 errors

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch committed:
updating spack completion

Signed-off-by: vsoch <vsochat@stanford.edu>

adamjstewart reviewed (approved):
This is great! I imagine a lot of other users will like this. Since it adds a new feature, let's get at least one more core developer to review. Also pinging some heavy Python users: @trws @skosukhin

scheibelp commented on lib/spack/spack/cmd/python.py:
```diff
@@ -48,10 +51,47 @@ def python(parser, args, unknown_args):
+        args.python_interpreter = "python"
+
+    # Run user choose of interpreter
+    if args.python_interpreter == "ipython":
```
If the user specifies `-i` and doesn't select `ipython` I think Spack should fail with an error message (in other words, I think there should be an `else` at the end of this branch that raises an exception if no expected identifier was provided).

Also, is there another interpreter besides `IPython` that could be used here (i.e. one available as a library)? If not, it might make sense to make `-i` a flag.

adamjstewart replied:
Argparse should handle that. The choices are `-i ipython` or `-i python` (the default).

vsoch replied:
@scheibelp there are indeed other interpreters - bpython as a quick example. In practice I'm not sure if it's likely a user would request these specific ones over ipython or python, but since it could very likely be a possibility, I think we should maintain the `-i` to maintain consistency of the UI and then allow for it. For example, if spack some day has some kind of interaction with a notebook (to quickly load and do some kind of quick data science thing with package metadata), the "interpreter" could very easily be something like jupyter.

scheibelp commented on lib/spack/spack/cmd/python.py:
```diff
@@ -48,10 +51,47 @@ def python(parser, args, unknown_args):
+    try:
+        import IPython
+    except ImportError:
+        tty.warning("ipython is not installed, using python.")
```
I think this should just fail if `IPython` cannot be imported, so no `try`/`except` is needed here.

vsoch replied:
How about we still catch the error, give the meaningful message, but then exit cleanly? If we just try to do the import and IPython isn't installed, the user will see the ugly traceback.

scheibelp commented on lib/spack/spack/cmd/python.py:
```diff
@@ -48,10 +51,47 @@ def python(parser, args, unknown_args):
 
+    # If arguments or a command are provided, just use python
+    if args.python_args or args.python_command:
+        args.python_interpreter = "python"
```
If the user specifies `-i` along with more arguments, or a `-c` option, I think an error should be raised.

vsoch replied:
Gotcha, this makes sense.

scheibelp reviewed (changes requested):
I have a few requests relating to default behavior.

Also, is there an IPython executable? I'm wondering if we could get more flexibility by specifying the executable as an argument, modifying e.g. `PYTHONPATH`, and then launching the executable (vs. importing the relevant interpreter library) - that would allow supporting other interpreter options without modifying Spack.

adamjstewart commented:
> Also, is there an IPython executable?

Yes, it's called `ipython`.

vsoch commented:
> Also, is there an IPython executable? I'm wondering if we could get more flexibility by specifying the executable as an argument, modifying e.g. PYTHONPATH, and then launching the executable (vs. importing the relevant interpreter library) - that would allow supporting other interpreter options without modifying Spack.

This would definitely make sense for something that we cannot interact with via Python, but for these interpreters that's not the case so I'm not sure it's warranted yet. Also, we would need to be able to still provide the local environment with spack ready to go, which would be more challenging allowing for the user to specify any old executable. It's a cool idea, but maybe we should consider it when there is evidence that users want to use their own custom interpreter executables?

vsoch committed:
updating interactive python

the client should exit if the user provides an argument/command paired with
anything other than python, as it is not really logical. We also exit if
ipython is not able to be imported, vs. we previously switched over to
python instead and issued a warning.

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch commented:
okay! I've updated the PR to more explicitly exit with an error message for the user when a combination of args is invalid (e.g., a command and anything other than ipython) or if ipython cannot be imported. See https://github.com/spack/spack/pull/20329/commits/a691e9ad8e265c787dedaa9ede610937a2970473.

scheibelp commented:
(EDIT: this was already addressed by https://github.com/spack/spack/pull/20329#discussion_r551601533)

>> Also, is there an IPython executable? I'm wondering if we could get more flexibility by specifying the executable as an argument, modifying e.g. PYTHONPATH, and then launching the executable (vs. importing the relevant interpreter library) - that would allow supporting other interpreter options without modifying Spack.

> This would definitely make sense for something that we cannot interact with via Python, but for these interpreters that's not the case so I'm not sure it's warranted yet

This makes sense but are there interpreters other than IPython or the default Python which are available as imports? If not, I think `-i` should be a flag option rather than a user-specified option. `-i` would be a bit too succinct though so IMO the flag should be called `--ipython`

vsoch commented:
Yes, all the ones that I mentioned are available to be called interactively. If it's a Python library that gets called from the command line, logically there is a way to bypass the client and do from within Python. Here are a few examples for the ones I mentioned:

 - [bpython.embed](https://github.com/bpython/bpython/blob/4e11d5ed6f4df3faab5446c4300f98cd564c8d9a/bpython/__init__.py#L39)
 - [jupyter](https://github.com/jupyter/notebook/issues/4873)

I think we should take this approach, the simplest one, until there is strong reason (user demand) to do otherwise.

trws commented:
This seems somewhat overfit to ipython to me personally, I think largely because of the fact the `spack python` command reuses the existing interpreter rather than executing a new one with the correct `PYTHONPATH` (which was on my list of things to try and fix actually).  This would be ok with me if it's documented that the only supported interpreters for now are something like "current" and "ipython", so we can expand it out to take a generic path later for things like pypy, iron python etc. later.

scheibelp commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -409,7 +426,10 @@ or a file:
 
-just like you would with the normal ``python`` command.
+just like you would with the normal ``python`` command. Note that for both
+commands and files, the default will always be the python interpreter,
```
With new edits, `-i ipython` will through an error so I think comment can be updated accordingly.

vsoch replied:
My bad, totally forgot about updating the docs! I'll do that now.

scheibelp reviewed (changes requested):
I think the only thing needed here is updating the documentation in line with the updated error behavior.

scheibelp commented:
> This would be ok with me if it's documented that the only supported interpreters for now are something like "current" and "ipython", so we can expand it out to take a generic path later for things like pypy, iron python etc. later.

@trws IMO the documentation added here was clear that only `ipython` was supported as an alternative. Is there a wording tweak you would suggest?

I also felt that this solution wouldn't get in the way of a set-`PYTHONPATH`-and-invoke-interpreter-executable-based solution, but let me know if you disagree.

vsoch committed:
updating user docs for python command

to explicitly state that the options are 1) current python (default)
or 2) ipython. And show that commands/args must be done with python

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch commented:
I updated the docs to be consistent with the current implementation, and (hopefully) did a better job to address @trws about explicitly stating the two options. https://github.com/spack/spack/pull/20329/commits/849b4c5e951c89a01b8003f86ac595a8cb3c3db5. I also used more consistent capitalization for Python and IPython.

-------------------

Strange! GitHub links the commit to the repository here (with a message that the commit doesn't belong) but it should be linked to my fork, https://github.com/vsoch/spack/commit/849b4c5e951c89a01b8003f86ac595a8cb3c3db5.

trws commented:
It probably won't get in the way.  The docs seem to have some remaining issues though, in that `spack python` still runs an interactive repl with or without the flag (and thus is most definitely interactive capable), and line 417 seems to imply that `spack python -i python` results in IPython?

vsoch committed:
removing confusing statement in docs

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch commented:
That was a typo, and having the redundancy is probably confusing. Fixed in https://github.com/spack/spack/pull/20329/commits/c4ca2190aa3dfd1b996f60a38f5d1ae20e7c78f6. The very first part of the section does mention that `spack python` is interactive, but for the script / command use case you wouldn't want it interactive, which is what I was trying to say, maybe I didn't do a good job? @trws is there a better way we could say this?

vsoch commented:
As a possible solution, I just removed the extra "clarification" (which wasn't very good, my bad!) about why we would only allow python for commands and scripts. It now just directly states the options without trying to explain them, which sounds much better I think.

scheibelp commented on lib/spack/spack/cmd/python.py:
```diff
@@ -48,10 +51,47 @@ def python(parser, args, unknown_args):
+    """An ipython interpreter is intended to be interactive, so it doesn't
+    support running a script or arguments
+    """
+    # If the user doesn't have ipython, fallback to python
```
I think this comment is out of date

vsoch replied:
Removed https://github.com/spack/spack/pull/20329/commits/a3f087f912df42eb62e5236dbdadb64874099aee

vsoch committed:
removing outdated comment

Signed-off-by: vsoch <vsochat@stanford.edu>

scheibelp commented on lib/spack/spack/cmd/python.py:
```diff
@@ -48,10 +51,46 @@ def python(parser, args, unknown_args):
+    if has_command and args.python_interpreter != "python":
+        tty.die("Use the default python interpreter for args/commands.")
+
+    # Run user choose of interpreter
```
Minor typo: choose -> choice

scheibelp commented on lib/spack/spack/cmd/python.py:
```diff
@@ -48,10 +51,46 @@ def python(parser, args, unknown_args):
+    # Run user choose of interpreter
+    if args.python_interpreter == "ipython":
+        return spack.cmd.python.ipython_interpreter(args)
+    return spack.cmd.python.python_interpreter(args)
```
My preference is that this statement is rearranged like

```
if python_interpreter == "ipython":
  ipython_interpreter(args)
elif python_interpreter == "python":
  python_interpreter(args)
else:
  raise SpackError("invalid choice of interpreter: {0}".format(python_interpreter))
```

adamjstewart replied:
The only problem with that is that it will be impossible to test the else clause since argparse prevents it from ever being reached

scheibelp replied:
Whoops I keep forgetting that. My bad.

vsoch committed:
choose -> choice

Signed-off-by: vsoch <vsochat@stanford.edu>

trws commented:
The doc updates look good. 👍 

The thing I'm contemplating now though is that while I occasionally use ipython, I almost never use it interactively.  I normally use it to get better errors, or turn on its pdb integration and use it as a debugger, things like this:

```
❯ ipython ~/scripts/broken-script.py
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/scripts/at-home.py in <module>
      8 host = check_output("hostname", shell=True)
      9
---> 10 if 'hurricane' in host or 'gale' in host:
     11     sys.exit(0)
     12 if "typhoon" in host or "storm" in host or "bolt" in host:

TypeError: a bytes-like object is required, not 'str'
```

Why prevent this use?

vsoch commented:
> The thing I'm contemplating now though is that while I occasionally use ipython, I almost never use it interactively. I normally use it to get better errors, or turn on its pdb integration and use it as a debugger, things like this:

Oh, interesting, I've never used it like that before! I typically just embed it everywhere to get an interactive session to develop and debug. @trws if I understand correctly, we should allow using ipython for the script use case (and string arguments as well?) and remove the checks that don't allow it?

trws commented:
If possible? I actually use embedding too, but usually by adding an `embed()` call into the script to break at that point.  I may just be strange like that, but if it wouldn't impose a significant amount of extra work it would be really nice, especially for debugging spack itself come to think of it.

vsoch committed:
adding support for ipython commands and scripts

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch commented:
@trws I gave it a shot! https://github.com/spack/spack/pull/20329/commits/8fa20641eefd1332a9b6d70d956319be277f8860. One thing I'm wondering about is if it should be allowed to run a command *and* a script. It seems to functionally work for python (and I tested for ipython but it's not included in the PR here) but I'm not sure it's working in practice. For example, let's say I have a script that prints the current working directory, but doesn't have the os import:

```python
# script.py
print(os.getcwd())
```
If I run the script with the python interpreter, this is what I'd expect:

```bash
$ bin/spack python script.py 
Traceback (most recent call last):
  File "script.py", line 2, in <module>
    print(os.getcwd())
NameError: name 'os' is not defined
```
Arguably since the console command is run before the script, I should be able to add it, and then see the printing:

```bash
$ bin/spack python -c "import os" script.py
# nothing printed
```
It's not an issue of the import working, because I can change to sys (and I'd expect to see the error message again) but I don't:

```bash
$ bin/spack python -c "import sys" script.py
# nothing printed
```
The only thing that "works" at least to run and still produce the output is to have the `-c` argument handed as a python arg instead of to the spack argparser:

```python
$ bin/spack python script.py -c "import os"
Traceback (most recent call last):
  File "script.py", line 2, in <module>
    print(os.getcwd())
NameError: name 'os' is not defined
```
But the import is still not done in context of the script.

So if this is a bug, while it's present we should check that both are defined in the parent function and cut out early if this is the case. If the user wants to run a command alongside a script, arguably it's just easiest to write it into the script. Let me know what you think!

trws commented:
Wow, I had not thought of that at all.  So, doing a bit of similar testing with python the executable, it seems as though if you specify `-c` then the positional arguments are treated as arguments to the command rather than executed as scripts, so I suppose I would do that?  Just looked at the documentation, it looks like `-c` terminates the argument list, so yeah probably simplify that way.  Really interesting question, and I like the thought of being able to inject something like that too, but for this probably go for consistency.

vsoch commented:
So just to clarify, you think that we shouldn’t allow specifying both (the spack -c, not a -c to go to Python) and exit early if this is the case? This is my suggestion because having both doesn’t seem to work! And of course the user is free to test specifying -c after a command so it’s handed off to Python or IPython.

vsoch committed:
do not allow the user to specify both

Signed-off-by: vsoch <vsochat@stanford.edu>

vsoch commented:
I just pushed a fix https://github.com/spack/spack/pull/20329/commits/75e5c8310e69d6bca731ff93d1a523977b12eaf6 to cut out early if both are specified. So this won't work:

```bash
$ spack python -c "import sys" script.py
==> Error: You can only specify a command OR script, but not both.
```
But this will, and the user is free to do whatever they like! Likely a separate command could be run first, independent of the script, but not something you expect to hand off to the script like an IPython embed.

```bash
$ bin/spack python script.py -c "import os"
Traceback (most recent call last):
  File "script.py", line 2, in <module>
    print(os.getcwd())
NameError: name 'os' is not defined
```

scheibelp commented:
Thanks!

vsoch commented:
Wooo merged! Thanks y'all! :D

trws commented:
Thank you, good addition, and thanks for working through the feedback.
